typed-transitions/src/Protocol/Chain/Type.hs

```diff
@@ -0,0 +1,116 @@
+  TrRespond (ResDownloadOne _) -> "ResDownloadOne"
+  TrRespond (ResDownloadDone _) -> "ResDownloadDone"
+  TrRespond _ -> "Res other"
+  TrConsumerDone -> "Consumer done"
+  TrProducerDone -> "Producer done"
+
+data RequestKind where
+  Improve  :: RequestKind
+  Download :: RequestKind
+  -- | Request for the next change in the chain.
+  Next     :: RequestKind
+
+data Request point (req :: RequestKind) where
+  -- | Request to update the read pointer to the newest of these blocks.
+  ReqImprove  :: NonEmpty point -> Request point 'Improve
+  -- | Request at most this many headers, from the current read pointer.
+  ReqDownload :: Word           -> Request point 'Download
+  -- | Request the next update to the producer's chain: a new tip (extension)
+  -- or a fork (read pointer is no longer in the producer's chain).
+  ReqNext     ::                   Request point 'Next
+
+-- | Responses parameterised by the request kind.
+data Response point header (req :: RequestKind) (res :: StChainExchange) where
+  -- | There was a fork: the read pointer has changed. This can be a response
+  -- to anything.
+  ResForked       :: (point, point)        -> Response point header anything 'StIdle
+  -- | The tip of chain has changed, but there was no fork (the new chain
+  -- includes the read pointer). This is only a response to 'ReqNext. If there
+  -- is an extension between request and response of other types, the new
+  -- tip is included in those responses.
+  ResExtend       :: point                 -> Response point header 'Next 'StIdle
+  -- | The new read pointer. If it's not in the requested set, then none of
+  -- them were in the chain.
+  ResImprove      :: (point, Maybe point)  -> Response point header 'Improve 'StIdle
+  -- | Multi-part download of headers.
+  ResDownloadOne  :: (header, Maybe point) -> Response point header 'Download ('StBusy 'Download)
+  -- | Multi-part download of headers finished.
+  ResDownloadDone :: Maybe point           -> Response point header 'Download 'StIdle
+
+-- |
+-- = Paritioning into client/server or consumer/producer
+--
+-- The above transition system 'TrChainExchange' makes no explicit mention
+-- of a client and a server. In theory, the protocol could be run by one
+-- process. In order to write separate client and server sides which are
+-- complementary, we must partition the states into 2 sets.
+
+-- | A type to identify our client/server parition.
+data ChainExchange
+
+type instance Partition ChainExchange st client server terminal =
+  ChainExchangePartition st client server terminal
+
+-- | Idle states are client, producer states are server.
+-- The init state is server, which is a bit weird. The server starts by giving
+-- its tip-of-chain, so that for the rest of the protocol, the consumer always
+-- knows the read pointer and the tip-of-chain.
+type family ChainExchangePartition st client server terminal where
+  ChainExchangePartition 'StInit       client server terminal = server
```

Comment:
It's a bit wired that the server starts the protocol.  I think it should be initiated by a client, which says "hey I want to follow your chain", this will fit nicer into a subscription model.

Reply:
> It's a bit wired that the server starts the protocol.

Yes indeed. So we should stop saying client and server :) It's completely fine that a _producer_ starts the protocol, isn't it?

Reply:
Sure, it doesn't matter.  And even for subscription, when the server receives a subscription request, it's his turn to reply, while in the other case the client needs to send a subscription and start the protocol (which might be just a single message).